src/HttpServer.gren

```diff
@@ -0,0 +1,332 @@
+    case s of
+        "GET" ->
+            GET
+        
+        "HEAD" ->
+            HEAD
+
+        "POST" ->
+            POST
+        
+        "PUT" ->
+            PUT
+        
+        "DELETE" ->
+            DELETE
+        
+        "CONNECT" ->
+            CONNECT
+        
+        "TRACE" ->
+            TRACE
+        
+        "PATCH" ->
+            PATCH
+        
+        _ ->
+            UNKNOWN
+
+
+arrayPairs : Array String -> Array (Array String)
+arrayPairs a =
+    let
+        pair = 
+            Array.takeFirst 2 a
+        
+        rest =
+            Array.dropFirst 2 a
+        
+        allPairs =
+            [ pair ] ++ case rest of
+                [] ->
+                    []
+                
+                _ ->
+                    arrayPairs rest
+    in
+    allPairs
+
+
+dictFromPairs : Array (Array String) -> Dict String String
+dictFromPairs pairs =
+    let
+        mapper p =
+            case p of
+                [a, b] ->
+                    Just { key = a, value = b }
+                
+                _ ->
+                    Nothing
+        
+        array =
+            Array.filterMap mapper pairs
+    in
+    Dict.fromArray array
+
+
+-- EFFECT STUFF
+
+
+type HttpCmd a
+    = SendResponse Response
+
+
+type HttpSub msg
+    = OnRequestSub Server (Request -> Response -> msg)
+
+
+cmdMap : (a -> b) -> HttpCmd a -> HttpCmd b
+cmdMap _ cmd =
+    case cmd of
+        SendResponse r ->
+            SendResponse r
+
+
+subMap : (a -> b) -> HttpSub a -> HttpSub b
+subMap f sub =
+    case sub of
+        OnRequestSub server msg ->
+            -- TODO: not sure this is right
+            OnRequestSub server (\req res -> f (msg req res))
+
+
+type alias State msg =
+    Array (HttpSub msg)
+
+
+init : Task Never (State msg)
+init =
+    Task.succeed []
+
+
+onEffects
+    : Platform.Router msg SelfMsg
+    -> Array (HttpCmd a)
+    -> Array (HttpSub msg)
+    -> State msg
+    -> Task Never (State msg)
+onEffects router cmds subs state =
+    let
+        -- TODO: look at Terminal.gren for a better way to manage subscriptions
+        -- For now just removing them all then re-adding.
+        _removeListeners =
+            state
+                |> Array.map
+                    (\(OnRequestSub server _) ->
+                        Gren.Kernel.HttpServer.removeAllListeners server
+                    )
+        
+        _addListeners =
+            subs
+                |> Array.map
+                    (\(OnRequestSub server msg) ->
+                        Gren.Kernel.HttpServer.addListener server router msg
+                    )
+
+        _sendResponses =
+            cmds
+                |> Array.map
+                    (\(SendResponse (Response kernelResponse response)) ->
+                        -- Terminal.gren sends to self and makes the kernel calls in onSelfMsg,
+                        -- but I couldn't get a self msg to trigger onSelfMsg when I tried that.
```

Comment:
FYI: I don't consider this to be a blocker for a release. Same applies to removing and re-adding listeners. If it all works as the user would expect it to, I'd prioritize these things last and instead focus on things required for a release :)

Reply:
And the reason for that, is that I expect these details to change for most effect managers in the future.